### PR TITLE
[MPS] Compile kernels into Metallib

### PR DESCRIPTION
**Pull request by malfet**
PyTorch MPS backend for the most part relies on MPSGraph to provide specific operations, but recently more and more often one had to implement custom kernel here that were simply embedded in the operator codebase and were compiled directly using [`- id<MTLLibrary>newLibraryWithSource:options:error:`](https://developer.apple.com/documentation/metal/mtldevice/1433431-newlibrarywithsource) (first metal kernel to MPS backend was added in https://github.com/pytorch/pytorch/pull/82307 )
Later on, as number of operator grew, those were refactored into `MetalShaderLibrary` convenience class (see  https://github.com/pytorch/pytorch/pull/125550 )

But as number of kernels keeps growing, it's time to make a next step and properly compile them into `.metalib`

This PR does exactly that by:
 - Moving shader sources into separate .metal files
 - Adds check on whether full Xcode installed or just DeveloperTools
 - If full Xcode is installed, compiles and links shaders into .metallib for Metal-3.0(Available on MacOS 13) and Metal-3.1 standard (available on MacOS 14, can use bfloat) and bundles both using `-sectcreate` linker option and `getsectiondata` API call. `metallib_dummy.cpp` file is used to properly express dependencies between metallib build and torch_cpu link stages. Logic for generating metallibraries is loosely based on https://github.com/ml-explore/mlx/blob/main/mlx/backend/metal/kernels/CMakeLists.txt. 
 - If only DeveloperTools CLI is installed, automatically wraps .metal into `_metallib.h` that contains shader source wrapped in `MetalShaderLibrary`

Bulk of changes introduced in this PR are just moving code around. I.e. for every file that contains non-templated shader definition in `aten/src/ATen/native/mps/operators` folder, corresponding `.metal` file is created in `aten/src/ATen/native/mps/kernels` folder and embedded shader definition is replaced with the following
```cpp
#ifndef PYTORCH_JIT_COMPILE_SHADERS
static auto& lib = MetalShaderLibrary::getBundledLibrary();
#else
#include <ATen/native/mps/OpName_metallib.h>
#endif
```

Some historical stats:
| PyTorch Version  | Number of shaders in MPS | Ops added |
| ------------- | ------------- | ---- |
| 1.12  | 0  | |
| 1.13  | 2  | bitwise_ops and  index.out |
| 2.0  | 4  | cross repeat and view)  |
| 2.1  | 9   | unary_ops, histogram, renorm, binary_ops |
| 2.2  | 11   | gamma and bucketization |
| 2.3  | 12  | naive_matmul (to workaround crash) |
| 2.4 | 13 | quantized_mm |
| 2.5 | 14 | fused_adam |

Pros:
  - Better code structure/readability
  - Eventually allows one to use shared headers (and implement something like `TensorIterator`)
  - Faster runtime (as compilation is done ahead of time) and perhaps better optimized compiled kernels

Cons:
  - Build process is a bit more complicated that it used to be
  - Need to maintain two codepath (as our CI builders only has DeveloperTools installed)

cc @albanD

### CODE REVIEW

**Comment by qqaatw** [comment]
Nice improvement! We rely on existing tests?

**Comment by malfet** [comment]
> Nice improvement! We rely on existing tests?

Yes, existing tests should be sufficient, though I would need to figure out how to install Xcode on our runners, otherwise they'll have to use embedded codepath

File: cmake/Metal.cmake [hunk 1]
```diff
@@ -2,6 +2,29 @@ if(NOT APPLE)
     return()
 endif()
 
+set(BFLOAT_METAL_CODE "
```
Comment (manuelcandales) [inline cmake/Metal.cmake:5]: Why do we have metal code in the .cmake file? Can me move that inside some utils.metal file?

**Review by manuelcandales — APPROVED** [review]

File: cmake/Metal.cmake (hunk 1 above)
Comment (malfet) [reply cmake/Metal.cmake:30]: This one is not a real kernels, it just checks if host can compile something, this is a pretty standard practice to embed trivial test code in `.cmake` files, see https://github.com/pytorch/pytorch/blob/86b45bde196b976247f2771196f8c1c689dd66b7/cmake/Modules/FindAVX.cmake#L5-L14 for example

**Comment by swolchok** [comment]
started getting build errors on my mac about missing aten/src/ATen/metallib_dummy.cpp locally on mac running `python setup.py develop`, even when running with `--cmake` or after `python setup.py clean`. working around via `USE_MPS=OFF python setup.py develop`

**Comment by wm901115nwpu** [comment]
> started getting build errors on my mac about missing aten/src/ATen/metallib_dummy.cpp locally on mac running `python setup.py develop`, even when running with `--cmake` or after `python setup.py clean`. working around via `USE_MPS=OFF python setup.py develop`

https://github.com/pytorch/pytorch/blob/3179eb15ae2ed08266392897ea18a498f91f8ba1/caffe2/CMakeLists.txt#L688

**Pull request closed without merging**